### DISCUSSION TITLE
Migrate alembic from monodb to multidb?

### DISCUSSION THREAD

**Discussion by Diego-Hernandez-Moodys** [sqlalchemy/alembic#1426] [Usage Questions] [answered]
We've already initialized our project using a template for a single database.

Is there any easy way for us to
1. Migrate our alembic.ini and migrations/env.py to use the multidb template
2. Retain the modification versions but point to a single database?

**Comment by CaselIT** [comment]
Hi,

I would init alembic using the multi-db template into a scratch directory and compare it with what you have, and apply the required changes to your current setup. They should not be too many

**Reply by Diego-Hernandez-Moodys** [reply]
I've made some progress and now I'm working on migrating the versions.
Is the only thing differentiating different databases in the multi-db is that the downgrade and update functions refer to the database explicitly?

e.g., old way
```python
"""Add official name

Revision ID: e34c2e65c6df
Revises: ef2f18a8fd24
Create Date: 2024-01-30 10:00:14.273803

"""
import alembic
import sqlalchemy

# revision identifiers, used by Alembic.
revision = 'e34c2e65c6df'
down_revision = 'ef2f18a8fd24'
branch_labels = None
depends_on = None

def upgrade():
    alembic.op.add_column(
        "org",
        sqlalchemy.Column("vendor_name", sqlalchemy.VARCHAR(256), nullable=True, insert_after="name") 
    )

def downgrade():
    alembic.op.drop_column("org", "vendor_name")
```

Now I just change the downgrade and upgrade functions to reflect the engine?
```python
"""Add official name

Revision ID: e34c2e65c6df
Revises: ef2f18a8fd24
Create Date: 2024-01-30 10:00:14.273803

"""
import alembic
import sqlalchemy

# revision identifiers, used by Alembic.
revision = 'e34c2e65c6df'
down_revision = 'ef2f18a8fd24'
branch_labels = None
depends_on = None

def upgrade_engine1():
    alembic.op.add_column(
        "org",
        sqlalchemy.Column("vendor_name", sqlalchemy.VARCHAR(256), nullable=True, insert_after="name") 
    )

def downgrade_engine1():
    alembic.op.drop_column("org", "vendor_name")
```

Do I still need these functions present?
```
def upgrade(engine_name):
    globals()["upgrade_%s" % engine_name]()


def downgrade(engine_name):
    globals()["downgrade_%s" % engine_name]()
```

**Reply by CaselIT** [reply]
these are created by the mako template used by alembic. The multidb one uses that:
https://github.com/sqlalchemy/alembic/blob/8829bd64a70708ef9144604ea0f0a5179d07ccbe/alembic/templates/multidb/script.py.mako#L24-L29

**Reply by Diego-Hernandez-Moodys** [reply]
Yes I saw that in the readme, but now I'm porting existing versions/*.py to the new format. What's minimally required?

**Reply by CaselIT** [reply, marked as answer]
in your case I think just renaming existing `upgrade` and `downgrade` to have the engine name you wanted, like you already shown, then adding to all migration the "standard" upgrade/downgrade from the multi template
```py
def upgrade(engine_name: str) -> None: 
     globals()["upgrade_%s" % engine_name]() 
  
  
def downgrade(engine_name: str) -> None: 
     globals()["downgrade_%s" % engine_name]() 
```

**Reply by Diego-Hernandez-Moodys** [reply]
That worked! Thank you.
One final note is the alembic_version in the database for the additional engine2 needs to have a row that points to head revision id, otherwise you'll get the
```
ERROR [alembic.util.messaging] Destination head is not a valid downgrade target from current head(s)
  FAILED: Destination head is not a valid downgrade target from current head(s)
```
error.
